### ISSUE TITLE
[BUG] Repeatedly displaying and hiding the popup at short intervals does not display the Popup as intended on iOS.

### ISSUE THREAD

**Issue by cat0363** [CommunityToolkit/Maui#1213]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Did you read the "Reporting a bug" section on Contributing file?

- [X] I have read the "Reporting a bug" section on Contributing file: https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#reporting-a-bug

### Current Behavior

This issue transfer from  from Discussion #1202.
Since I was able to write code that can easily reproduce the problem, I uploaded the source code to github.

I am using Popup to display the indicator. Indicators using Popup are displayed before the
start of processing and closed after the end of processing. The length of the processing varies,
some long, some short. 

The Indicator using the Popup is displayed as follows.

    private void ShowIndicator()
    {
        if (pIndicator == null)
        {
            pIndicator = new Indicator();
            this.ShowPopup(pIndicator);
        }
    }

The Indicator using the Popup is closed as follows.

    private void CloseIndicator()
    {
        if (pIndicator != null)
        {
            pIndicator.Close();
            pIndicator = null;
        }
    }

pIndicator is an instance of Indicator class that inherits from Popup class and is a class variable.
Popups are managed by class variables, and only one Popup is displayed at a time.

    private Indicator pIndicator = null;

I use it as follows.

    ShowIndicator();
    GetTestItem();
    CloseIndicator();

If there is no problem, it will be displayed as follows.

![Capture_001](https://github.com/CommunityToolkit/Maui/assets/125236133/acb1d54d-4011-485e-a74c-fd66448304a4)

If the following warning is displayed, it will be displayed as follows.

Warning: Attempt to dismiss from view controller <CommunityToolkit_Maui_Core_Views_MauiPopup: 0x16c8d4250> while a presentation or dismiss is in progress!

![Capture_002](https://github.com/CommunityToolkit/Maui/assets/125236133/fe30b12c-2edb-4ccf-8f59-3b65be84ea0c)

Below is a video of the issue being reproduced.

https://github.com/CommunityToolkit/Maui/assets/125236133/6acfcb64-e1be-4fb8-9657-854d563d2b12

A Indicator using an Popup will appear in a position other than the center of the screen and will not close at all.
The above phenomenon rarely occurs.

As far as I can see from the warnings, it looks like the object was not destroyed immediately after calling the Popup's Close method and is still there.

The problem occurs in both of the repros below. 
The former is the reproduction code described in Discussion #1202.
The latter is the reproducible code described in this issue.

https://github.com/cat0363/MauiComm-IssuePopup.git
https://github.com/cat0363/MauiComm-IssuePopup2.git

### Expected Behavior

I expected the indicator to show/hide even after calling ShowPopup and Close in succession.

### Steps To Reproduce

The steps to reproduce are as follows.

1. Launch the app uploaded to github with the device on iOS.
2. Scrolls the ScrollView downward many times.

In step 2, the Indicator remains displayed at the bottom of the screen, blocking scrolling.

### Link to public reproduction project repository

https://github.com/cat0363/MauiComm-IssuePopup2.git

### Environment

```markdown
- .NET MAUI CommunityToolkit:5.2.0
- OS:iOS 16.4
- .NET MAUI:7.0.86
```

**Comment by cat0363** [comment]
By changing the Community Toolkit code as follows, the problem was resolved, 
but the warning is still output.

[MauiPopup.macios.cs]

    void AddToCurrentPageViewController(UIViewController viewController)
    {
        if (!viewController.IsBeingPresented)
        {
            viewController.PresentViewController(this, true, null);
        }
    }

[PopupHandler.macios.cs]

    public static async void MapOnClosed(PopupHandler handler, IPopup view, object? result)
    {
        var vc = handler.PlatformView.ViewController;
        if (vc is not null)
        {
            if (!vc.IsBeingDismissed)
            {
                await vc.DismissViewControllerAsync(true);
            }
        }

        handler.DisconnectHandler(handler.PlatformView);
    }


https://github.com/CommunityToolkit/Maui/assets/125236133/a48dd9b7-6afd-4f27-affa-66d1496f9bf9

Changed to check the value of IsBeginPresented property and IsBeginDismissed 
property before calling PresentViewController method and DismissViewControllerAsync 
method. I don't know the impact on others, so could you make a judgment?

**Comment by bijington** [comment]
It might be worth looking over what @brminnick is working on in this PR:

https://github.com/CommunityToolkit/Maui/pull/1223

**Comment by cat0363** [comment]
Hi, @bijington 
Thank you for the information.
I'll check the changes in that PR.

**Comment by bijington** [comment]
@cat0363 thanks for taking a look. Sorry I didn't quite follow, are you saying that when we merge #1223 this issue will be resolved?

**Comment by cat0363** [comment]
Hi, @bijington 
Thank you for your reply. Merging #1223 does not solve this problem.
In addition to #1223, I need the solution I posted earlier.

https://github.com/CommunityToolkit/Maui/issues/1213#issuecomment-1580300733

Can anyone help me determine that I need to check the values ​​of the IsBeingPresented 
and IsBeingDismissed properties before calling the PresentViewController and 
DismissViewControllerAsync methods?

I used the following as a reference when considering a solution.
https://stackoverflow.com/questions/12261008/strange-warning-dismissing-modal-view-controller

**Comment by cat0363** [comment]
Additional Information:
I mentioned that applying [my solution](https://github.com/CommunityToolkit/Maui/issues/1213#issuecomment-1580300733) will output another Warning, but it seems to be output on the .NET MAUI side.

    Warning: Attempt to dismiss from view controller <Microsoft_Maui_Controls_Platform_Compatibility_ShellFlyoutRenderer: 0x10313db20> while a presentation or dismiss is in progress!

The warning that was output before applying the solution is as follows.

    Warning: Attempt to dismiss from view controller <CommunityToolkit_Maui_Core_Views_MauiPopup: 0x11ddb6b90> while a presentation or dismiss is in progress!

By solving the problem, the warning that was output on the .NET MAUI side is now output.
Perhaps this seems like a similar issue.

**Comment by cat0363** [comment]
Hi, @brminnick 
If possible, I would like to hear your opinion.
When this issue is resolved, all the issues I want to resolve about Popup will be resolved.

**Comment by cat0363** [comment]
Hi, @brminnick 
By solving Issue #1111 with PR #1223, which you are currently working on, this issue will also be resolved.
I was able to confirm that using PR #1223 fixed the issue. This issue was resolved by awaiting using the 
CloseAsync method. Once #1222 is closed, I will close this issue. Popup behavior is now what I expected. 
Thank you.

It turns out that the [solution](https://github.com/CommunityToolkit/Maui/issues/1213#issuecomment-1580300733) I posted earlier only guarded the occurrence of the problem, not the root solution. 
By controlling the calls to the PresentViewController and DismissViewControllerAsync methods using the 
ViewController's IsBeingPresented and IsBeingDismissed property values, the phenomenon no longer 
occurs, but when I checked the Popup display count, it was not the intended count.
